src/ott/solvers/nn/neuraldual.py

```diff
@@ -96,66 +153,195 @@ def __init__(
+        f"the `W2NeuralDual` setting, with positive weights " \
+        f"being {self.pos_weights}."
+    if isinstance(
+        neural_f, icnn.ICNN
+    ) and neural_f.pos_weights is not self.pos_weights:
+      warnings.warn(warn_str)
       neural_f.pos_weights = self.pos_weights
+
+    if isinstance(
+        neural_g, icnn.ICNN
+    ) and neural_g.pos_weights is not self.pos_weights:
+      warnings.warn(warn_str)
       neural_g.pos_weights = self.pos_weights
 
-    self.state_f = neural_f.create_train_state(rng_f, optimizer_f, input_dim)
-    self.state_g = neural_g.create_train_state(rng_g, optimizer_g, input_dim)
+    self.state_f = neural_f.create_train_state(
+        rng_f, optimizer_f, input_dim, init_f_params
+    )
+    self.state_g = neural_g.create_train_state(
+        rng_g, optimizer_g, input_dim, init_g_params
+    )
 
-    # define train and valid step functions
-    self.train_step_f = self.get_step_fn(train=True, to_optimize="f")
-    self.valid_step_f = self.get_step_fn(train=False, to_optimize="f")
+    # Assume g defines the potential unless this attribute is set.
+    self.g_returns_potential = getattr(neural_g, 'returns_potential', True)
+
+    if not self.g_returns_potential and self.back_and_forth:
+      raise ValueError(
+          'back_and_forth not supported when g is the gradient of the potential'
+      )
 
-    self.train_step_g = self.get_step_fn(train=True, to_optimize="g")
-    self.valid_step_g = self.get_step_fn(train=False, to_optimize="g")
+    # default to using back_and_forth unless g provides the gradient
+    if self.back_and_forth is None:
+      self.back_and_forth = self.g_returns_potential
+
+    if self.num_inner_iters == 1 and self.parallel_updates:
+      self.train_step_parallel = self.get_step_fn(
+          train=True, to_optimize="both"
+      )
+      self.valid_step_parallel = self.get_step_fn(
+          train=False, to_optimize="both"
+      )
+      self.train_fn = self.train_neuraldual_parallel
+    else:
+      if self.parallel_updates:
+        warnings.warn(
+            'parallel_updates set to True but disabling it because num_inner_iters>1'
+        )
+      if self.back_and_forth:
+        raise NotImplementedError(
+            "back_and_forth not implemented without parallel updates"
+        )
+      self.train_step_f = self.get_step_fn(train=True, to_optimize="f")
+      self.valid_step_f = self.get_step_fn(train=False, to_optimize="f")
+      self.train_step_g = self.get_step_fn(train=True, to_optimize="g")
+      self.valid_step_g = self.get_step_fn(train=False, to_optimize="g")
+      self.train_fn = self.train_neuraldual_alternating
 
   def __call__(
       self,
       trainloader_source: Iterable[jnp.ndarray],
       trainloader_target: Iterable[jnp.ndarray],
       validloader_source: Iterable[jnp.ndarray],
       validloader_target: Iterable[jnp.ndarray],
+      callback: Optional[Callback_t] = None,
```

Comment:
Missing type hints of the callback.

Reply:
It's hinting to `Optional[Callback_t]`, is there something else I should add?